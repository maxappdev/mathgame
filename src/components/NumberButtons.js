import React from 'react';
import { View} from "react-native"
import NumberButtonsRow from './NumberButtonsRow'

const NumberButtons = (props) => {
    return (
        <>
        <View>
            <NumberButtonsRow handleButtonPress={props.handleButtonPress} buttonNames={["1","2","3"]}></NumberButtonsRow>
            <NumberButtonsRow handleButtonPress={props.handleButtonPress} buttonNames={["4","5","6"]}></NumberButtonsRow>
            <NumberButtonsRow handleButtonPress={props.handleButtonPress} buttonNames={["7","8","9"]}></NumberButtonsRow>
            <NumberButtonsRow handleButtonPress={props.handleButtonPress} buttonNames={["00","0","C"]}></NumberButtonsRow>
        </View>
        </>
    );
};
export default NumberButtons