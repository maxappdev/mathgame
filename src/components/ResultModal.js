import React from 'react';
import { Modal, View, Text, Button } from 'react-native';
import styles from '../styles/ResultModalStyles'

const ResultModal = (props) => {
    return (
    <>
    <Modal transparent={false} visible={props.isVisible} animationType="slide">
        <View style={styles.modalView}>
            <View>
            <Text>Your result: {props.points} POINTS</Text>
            </View>
            <View style={styles.buttonContainer}>
                <Button title="NEW GAME" onPress={props.newGame} />
                <Button title="TO MENU" onPress={props.backToMenu}/>
            </View>
            <View style={styles.buttonContinueContainer}>
                <Button title="CONTINUE GAME (SEE ADD)" onPress={props.continueGame} />
            </View>
        </View>
    </Modal>
    </>
    );
}
export default ResultModal