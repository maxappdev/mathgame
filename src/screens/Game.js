import React from 'react';
import {
  View,
  Text,
} from 'react-native';
import {
  useState,
  useEffect,
} from 'react'
import NumberButtons from '../components/NumberButtons'
import {useInterval} from '../logic/services/TimerService'
import {generateEquation} from '../logic/factories/EquationFactory'
import {solveEquation} from '../logic/services/EquationService'
import appStyles from '../styles/AppStyles'
import ResultModal from '../components/ResultModal'
import Level from '../logic/models/Level';
import {pageMountedFirstTime} from '../logic/services/PageRenderService'
import { maxLevelNr } from '../../src/logic/models/Level'

const Game = (props) => {

  const timerDelayInMs = 1000
  const timerDelayInSec = timerDelayInMs/1000
  const maxAnswerLength = 10
  const nextLevelAfter = 5
  const mountedFirstTime = pageMountedFirstTime();

  const [points, setPoints] = useState(0) 
  const [levelNr, setLevelNr] = useState(1)
  const [timerDurationInSec, setTimerDuration] = useState(Level.levelNumber[levelNr].duration)
  const [equation, setEquation] = useState(generateEquation(levelNr))
  const [answer, setAnswer] = useState("")
  const [timeoutVal, setTimeoutVal] = useState(timerDurationInSec)
  const [timerRunning, setTimerRunning] = useState(true)
  const [resultModalVisible, setResultModalVisible] = useState(false)

  useInterval(() => {
    setTimeoutVal(timeoutVal - timerDelayInSec);
    if(timeoutVal == 0){
      gameLost()
    }
  }, timerRunning ? timerDelayInMs : null);

  useEffect(() => {
    if (!mountedFirstTime) {
      checkAnswer()
    }
  },[answer]);

  useEffect(() => {
    if (!mountedFirstTime) {
      checkPoints()
    }
  },[points]);

  useEffect(() => {
    if (!mountedFirstTime) {
      setEquation(generateEquation(levelNr))
    }
  },[levelNr]);

  useEffect(() => {
    if (!mountedFirstTime) {
      const levelDuration = Level.levelNumber[levelNr].duration
      setTimerDuration(levelDuration)
      setTimeoutVal(levelDuration)
    }
  },[equation]);

  const checkAnswer = () => {
    if(answer != "" && answer == solveEquation(equation)){
      setAnswer("")
      setPoints(points+1)
    }
  }

  const checkPoints = () => {
    if(points == 0){
      setLevelNr(1)
    }
    else if(points != 0 && points%nextLevelAfter == 0 && ((levelNr+1) <= maxLevelNr ) ){
      setLevelNr(levelNr+1)
    }
    else{
      setEquation(generateEquation(levelNr))
    }
  }
  
  const gameLost = () => {
    setTimerRunning(false)
    setResultModalVisible(true)
  }

  const backToMenu = () => {props.navigation.navigate('Home')}

  const newGame = () => {
    setPoints(0)
    setAnswer("")
    setTimerRunning(true)
    setResultModalVisible(false)
  }

  const continueGame = () => {
    setEquation(generateEquation(levelNr))
    setAnswer("")
    setTimerRunning(true)
    setResultModalVisible(false)
  }

  const addToAnswer = (symbol) => {
    if( (answer.length < maxAnswerLength) || (symbol == "C") ){
      (symbol == "C") ? setAnswer("") : setAnswer(answer + symbol)
    }
  }
  
  return (
    <>
      <View>
        <ResultModal backToMenu={backToMenu} continueGame={continueGame} newGame={newGame} points={points} isVisible={resultModalVisible}></ResultModal>
        <View style={
        {
          width: timeoutVal*(100/timerDurationInSec)+'%',
          height: 30,
          backgroundColor: 'red',
          justifyContent: 'center',
          alignItems: 'center'
        }
        }>
        </View>
        <View style={appStyles.infoTextView}>
          <Text>POINTS: {points} ; LEVEL: {levelNr}</Text>
        </View>
        <View id="equation-view" style={appStyles.equationView}>
        <Text id="equation" style= {appStyles.equation}>{equation.print()}</Text>
        </View>
        <View id="answer-view" style={appStyles.answerView}>
          <Text id="answer">{answer}</Text>
        </View>
        <NumberButtons handleButtonPress={addToAnswer}></NumberButtons>
      </View>
    </>
  );
};

export default Game;