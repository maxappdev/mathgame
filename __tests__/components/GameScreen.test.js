import 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';
import ResultModal from '../../src/components/ResultModal';
import NumberButtons from '../../src/components/NumberButtons';
import { shallow } from 'enzyme';
import Game from '../../src/screens/Game';

describe('Game Screen', () => {
  it('renders Modal correctly', () => {
    const wrapper = renderer.create(<ResultModal />).toJSON()
    expect(wrapper).toMatchSnapshot();
  });

  it('renders Buttons Row correctly', () => {
    const wrapper = renderer.create(<NumberButtons />).toJSON()
    expect(wrapper).toMatchSnapshot();
  });

  it('contains equation', () => {
      const wrapper = shallow(<Game />)
      const equation = wrapper.find('View').find('#equation-view').find('#equation').render().text()
      expect(equation).toEqual(expect.any(String));  
  })

})

//TODO: write this method. it does not work and will be really difficult to write
/*
const userPlaysGameAndEarns100Points = () => {
    let gameScreen = shallow(<Game />) -> enzyme shallow method
    let equation = gameScreen.find('View').find('#equation-view').find('#equation').render().text()
    console.log(equation)
    let usersAnswer = getAnswer(equation){
        const getAnswer = (equationString) => {
    const properString = equationString.split("ร").join("*").split("รท").join("/")
    return eval(properString)
}
    }
     //TODO: now i just hardcode input. I should simulate pressing on real buttons
    const inputAnswer = gameScreen.find('View').find('NumberButtons').props().handleButtonPress
    inputAnswer(inputAnswer)
    shallow(gameScreen.find('#equation-view').find('#equation')).update()
    gameScreen = shallow(<Game />)
    equation = gameScreen.find('View').find('#equation-view').find('#equation').render().text()
    console.log(equation)
}
*/

//TODO: make it ready
/*describe('User on Game Screen', () => {
    it('can play game and earn 100 points', () => {
        userPlaysGameAndEarns100Points()
    })
})*/