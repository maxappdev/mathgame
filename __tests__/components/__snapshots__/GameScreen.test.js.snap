// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game Screen renders Buttons Row correctly 1`] = `
<View>
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <View
      style={
        Object {
          "height": 70,
          "width": "33%",
        }
      }
    >
      <View
        accessibilityRole="button"
        accessibilityState={Object {}}
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Array [
              Object {},
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#007AFF",
                  "fontSize": 18,
                  "margin": 8,
                  "textAlign": "center",
                },
              ]
            }
          >
            1
          </Text>
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "height": 70,
          "width": "33%",
        }
      }
    >
      <View
        accessibilityRole="button"
        accessibilityState={Object {}}
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Array [
              Object {},
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#007AFF",
                  "fontSize": 18,
                  "margin": 8,
                  "textAlign": "center",
                },
              ]
            }
          >
            2
          </Text>
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "height": 70,
          "width": "33%",
        }
      }
    >
      <View
        accessibilityRole="button"
        accessibilityState={Object {}}
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Array [
              Object {},
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#007AFF",
                  "fontSize": 18,
                  "margin": 8,
                  "textAlign": "center",
                },
              ]
            }
          >
            3
          </Text>
        </View>
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <View
      style={
        Object {
          "height": 70,
          "width": "33%",
        }
      }
    >
      <View
        accessibilityRole="button"
        accessibilityState={Object {}}
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Array [
              Object {},
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#007AFF",
                  "fontSize": 18,
                  "margin": 8,
                  "textAlign": "center",
                },
              ]
            }
          >
            4
          </Text>
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "height": 70,
          "width": "33%",
        }
      }
    >
      <View
        accessibilityRole="button"
        accessibilityState={Object {}}
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Array [
              Object {},
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#007AFF",
                  "fontSize": 18,
                  "margin": 8,
                  "textAlign": "center",
                },
              ]
            }
          >
            5
          </Text>
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "height": 70,
          "width": "33%",
        }
      }
    >
      <View
        accessibilityRole="button"
        accessibilityState={Object {}}
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Array [
              Object {},
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#007AFF",
                  "fontSize": 18,
                  "margin": 8,
                  "textAlign": "center",
                },
              ]
            }
          >
            6
          </Text>
        </View>
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <View
      style={
        Object {
          "height": 70,
          "width": "33%",
        }
      }
    >
      <View
        accessibilityRole="button"
        accessibilityState={Object {}}
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Array [
              Object {},
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#007AFF",
                  "fontSize": 18,
                  "margin": 8,
                  "textAlign": "center",
                },
              ]
            }
          >
            7
          </Text>
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "height": 70,
          "width": "33%",
        }
      }
    >
      <View
        accessibilityRole="button"
        accessibilityState={Object {}}
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Array [
              Object {},
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#007AFF",
                  "fontSize": 18,
                  "margin": 8,
                  "textAlign": "center",
                },
              ]
            }
          >
            8
          </Text>
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "height": 70,
          "width": "33%",
        }
      }
    >
      <View
        accessibilityRole="button"
        accessibilityState={Object {}}
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Array [
              Object {},
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#007AFF",
                  "fontSize": 18,
                  "margin": 8,
                  "textAlign": "center",
                },
              ]
            }
          >
            9
          </Text>
        </View>
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <View
      style={
        Object {
          "height": 70,
          "width": "33%",
        }
      }
    >
      <View
        accessibilityRole="button"
        accessibilityState={Object {}}
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Array [
              Object {},
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#007AFF",
                  "fontSize": 18,
                  "margin": 8,
                  "textAlign": "center",
                },
              ]
            }
          >
            00
          </Text>
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "height": 70,
          "width": "33%",
        }
      }
    >
      <View
        accessibilityRole="button"
        accessibilityState={Object {}}
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Array [
              Object {},
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#007AFF",
                  "fontSize": 18,
                  "margin": 8,
                  "textAlign": "center",
                },
              ]
            }
          >
            0
          </Text>
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "height": 70,
          "width": "33%",
        }
      }
    >
      <View
        accessibilityRole="button"
        accessibilityState={Object {}}
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Array [
              Object {},
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#007AFF",
                  "fontSize": 18,
                  "margin": 8,
                  "textAlign": "center",
                },
              ]
            }
          >
            C
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;

exports[`Game Screen renders Modal correctly 1`] = `
<Modal
  animationType="slide"
  hardwareAccelerated={false}
  transparent={false}
  visible={true}
>
  <View
    style={
      Object {
        "alignItems": "center",
        "height": "100%",
        "justifyContent": "center",
      }
    }
  >
    <View>
      <Text>
        Your result: 
         POINTS
      </Text>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "space-between",
          "marginTop": 20,
          "width": "60%",
        }
      }
    >
      <View
        accessibilityRole="button"
        accessibilityState={Object {}}
        accessible={true}
        focusable={false}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Array [
              Object {},
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#007AFF",
                  "fontSize": 18,
                  "margin": 8,
                  "textAlign": "center",
                },
              ]
            }
          >
            NEW GAME
          </Text>
        </View>
      </View>
      <View
        accessibilityRole="button"
        accessibilityState={Object {}}
        accessible={true}
        focusable={false}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Array [
              Object {},
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#007AFF",
                  "fontSize": 18,
                  "margin": 8,
                  "textAlign": "center",
                },
              ]
            }
          >
            TO MENU
          </Text>
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "justifyContent": "center",
          "marginTop": 20,
        }
      }
    >
      <View
        accessibilityRole="button"
        accessibilityState={Object {}}
        accessible={true}
        focusable={false}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Array [
              Object {},
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#007AFF",
                  "fontSize": 18,
                  "margin": 8,
                  "textAlign": "center",
                },
              ]
            }
          >
            CONTINUE GAME (SEE ADD)
          </Text>
        </View>
      </View>
    </View>
  </View>
</Modal>
`;
